### ISSUE TITLE
Question/feature request: Stripe splitting/metadata optimization for older data arriving later  

### ISSUE THREAD

**Issue by muntdan** [hydradatabase/hydra#162]
### What's wrong?

~~FIY: Discord channel is not working and~~ you allow now only bug reports but this is are questions/feature request

If I have data arriving later (data for 2 days ago) and I run a ETL every 15 min towards Hydra:
Should I at least insert the older data separate than data from the last 15 min or should I do something else ? 
Will it find the existing chunk for that day and append there ?
Should I worry about data timespan fragmentation/overlapping between different stripe/chunks ?

**Comment by wuputah** [comment]
Hydra does not "cluster" data ([in the Postgres sense](https://www.postgresql.org/docs/current/sql-cluster.html)) at this time, the data would be appended at the end of the table in a new stripe. This will work fine and isn't inefficient, but if this data gets "mixed in" with older data, the metadata would be less specific for that one stripe, so it might get pulled into 

The way I see it, the feature you want is one you requested already, #62 but maybe you see these as different things?

**Comment by muntdan** [comment]
Indeed I see this as a different feature, not related to the partitioning of data, but the out of order arriving data scenario => stripe with wide metadata instead of narrow/efficient.

**Comment by wuputah** [comment]
In order to do that in a reasonable way, every stripe would need to be "appendable", but that's not the way the storage works currently. Currently data is always appended to the end of the table, because the table is one big file, there's no place to add data to previously written stripes. The only alternative way to do that it would need to delete  the entire stripe as deleted and rewrite it at the end, which would be very expensive (e.g. insert 1 row could mean rewrite 150,000 rows of data). It would also make a "hole" in the file that then needs to be filled back in later (which is actually what `columnar.vacuum` does - it rewrites stripes that has deleted data, and tries to move any new stripes it creates into any existing holes made by its own process).

The database would also needs a way to decide what is the most important metadata by which to group data, i.e. how does it find the stripe which is the most optimal to put the data in. This is, basically, defining a clustering or partitioning key.

So, basically: built-in, behind the scenes, auto-partitioning by a particular column is just data clustering by a different name.

Anyway it's all something we are thinking about for the future, a big part of what we are doing now is learning the limitations of the implementation and what we need to change/improve.

**Comment by muntdan** [comment]
I understand now why you were thinking of partitioning issue for this case also.
I will try to configure and test on Hydra also 2 dimension partition: tenant and months to see how it goes:
https://stackoverflow.com/questions/64207256/how-to-create-multiple-column-partitions-for-postgresql-db-one-for-time-range-a
If I encounter any issues will update/transform this issue into a bug in regards to manual partitioning 2 dimension on Hydra.

**Comment by wuputah** [comment]
Sorry, I don't know (and couldn't find) what Postgres does if you manually run `vacuum` on a partitioned table, but this could be tested. 

A few ideas:

* You could `vacuum` without argument to run vacuum on all tables.
* You could try [tweaking your autovacuum settings](https://www.postgresql.org/docs/current/runtime-config-autovacuum.html) to run vacuum at the frequency you need. For example, you could set `autovacuum_vacuum_insert_threshold` to `150000` and `autovacuum_vacuum_insert_scale_factor` to `0` to try to trigger autovacuum every 150k rows, but this would affect the entire server (including `heap` tables). 
* You could modify your autovacuum settings on a per-table basis. The documentation states:
  
  > the setting can be overridden for individual tables by changing table storage parameters
  
  These can set using the `WITH` clause to [CREATE TABLE](https://www.postgresql.org/docs/current/sql-createtable.html#SQL-CREATETABLE-PARMS-WITH), or `SET` to [ALTER TABLE](https://www.postgresql.org/docs/16/sql-altertable.html#SQL-ALTERTABLE-DESC-SET-STORAGE-PARAMETER). See [storage parameters](https://www.postgresql.org/docs/current/sql-createtable.html#SQL-CREATETABLE-STORAGE-PARAMETERS) for more information about the settings that can be set this way.

**Comment by muntdan** [comment]
I tested now and is great news vacuum on top parent table seems to be propagated all the way down:
![image](https://github.com/hydradatabase/hydra/assets/105269340/8acd1833-7da5-4bab-83b8-0465496552cb)

But unfortunately I see a degradation in query time: 
For 15 cars X 30days | 1 column retrieved speed with all the max config you have for Clickbench https://github.com/hydradatabase/ClickBench/blob/bf41de4d6e8a3977c91a434915704d31be9a1447/hydra/install.sh
the duration increase from 1.3sec with only 10 Tenant partitions to 2.4sec now with secondary weekly partitions also.

**Comment by wuputah** [comment]
ah nice.

> the duration increase from 1.3sec with only 10 Tenant partitions to 2.4sec now with secondary weekly partitions also.

I would think this depends on your query, the more tables your query matches (i.e. how many weeks / tenants) it will go slower than having it in one table, but if you're only querying a subset of the tables, I would think it would be faster. So, ideally, you should match the partitions to your query patterns.
